metaflow/datatools/s3op.py

```diff
@@ -73,17 +92,73 @@ def worker(queue, mode):
             if url is None:
                 break
 
-            if mode == 'download':
+            if mode == 'info':
+                try:
+                    head = s3.head_object(Bucket=url.bucket, Key=url.path)
+                    to_output = {
+                        'error': None,
+                        'size': head['ContentLength'],
+                        'content_type': head['ContentType'],
+                        'metadata': head['Metadata']}
+                except ClientError as err:
+                    error_code = normalize_client_error(err)
+                    if error_code == 404:
+                        to_output = {'error': ERROR_URL_NOT_FOUND}
+                    elif error_code == 403:
+                        to_output = {'error': ERROR_URL_ACCESS_DENIED}
+                    else:
+                        to_output = {'error': error_code}
+                        raise
+                finally:
+                    with open(url.local, 'w') as f:
+                        json.dump(to_output, f)
+            elif mode == 'download':
                 tmp = NamedTemporaryFile(dir='.', delete=False)
                 try:
-                    s3.download_file(url.bucket, url.path, tmp.name)
+                    if url.range is None:
+                        s3.download_file(url.bucket, url.path, tmp.name)
+                    else:
+                        # We do get_object. We don't actually do any retries
+                        # here because the higher levels will do the retry if
+                        # needed
+                        resp = s3.get_object(
+                            Bucket=url.bucket,
+                            Key=url.path,
+                            Range=url.range)
+                        code = str(resp['ResponseMetadata']['HTTPStatusCode'])
+                        if code[0] == '2':
+                            tmp.write(resp['Body'].read())
+                        else:
+                            tmp.close()
```

Comment:
aren't `tmp.close()` and `os.unlink` redundant given `except` below?

Reply:
Absolutely; will remove.

Reply:
Done